### PR TITLE
MRG: improve parallel backends

### LINKED ISSUE #210: MPIBackend does not allow passing whole objects
Currently, the MPIBackend passes the `params` dictionary to the child for running simulations. I think we should allow passing arbitrary Python objects as long as they are picklable. This will make it more similar to the Joblibs backend and also not limit code design. @blakecaldwell do you have an intuition whether this is easy/difficult to implement.

[Discovered this](https://github.com/jonescompneurolab/hnn-core/pull/199#discussion_r528268228) in #199 with @ntolley

### PR DESCRIPTION
**Pull request by jasmainak**
closes #210 

<s>This is mostly a PR stating my intent so far</s>

cc @blakecaldwell @ntolley @cjayb 

will ping again when ready, also attempting to improve the "bus factor" by messing with the code (not that we expect anything to happen to @blakecaldwell :)

### CODE REVIEW

File: hnn_core/parallel_backends.py
Comment (jasmainak) [inline hnn_core/parallel_backends.py:342]: not quite sure why this was necessary?

Comment (blakecaldwell) [reply hnn_core/parallel_backends.py:342]: JoblibBackend for 1 job is essentially no backend. Don't bother creating the mpi child process for a single processor.

File: doc/parallel.rst
Comment (jasmainak) [inline doc/parallel.rst:104]: just broke down the lines because it makes it easier to read.

**Comment by jasmainak** [comment]
ready for review. @ntolley now we pass the `Network` object instead of `net.params` so both backends are equivalent

File: hnn_core/parallel_backends.py
Comment (jasmainak) [inline hnn_core/parallel_backends.py:227]: explicitly using the name of the functions to make the code a bit more transparent

File: hnn_core/parallel_backends.py
Comment (jasmainak) [inline hnn_core/parallel_backends.py:159]: tried to make something more generic. Didn't completely succeed, will try more in future PRs

File: hnn_core/parallel_backends.py
Comment (jasmainak) [inline hnn_core/parallel_backends.py:506]: one of the main objectives of this redesign was to have as little code in `MPIBackend.simulate()` as in `JoblibBackend.simulate()`

**Comment by jasmainak** [comment]
anyone got a chance to look? @blakecaldwell perhaps you're the best person to review this PR?

File: hnn_core/parallel_backends.py
Comment (ntolley) [inline hnn_core/parallel_backends.py:68]: Looks like these issues are dependent on the ompi version. Are there other issues that would necessitate **ompi>=ver** in the install instructions?

File: hnn_core/parallel_backends.py
Comment (ntolley) [inline hnn_core/parallel_backends.py:393]: list of str?

File: hnn_core/parallel_backends.py
Comment (ntolley) [inline hnn_core/parallel_backends.py:504]: Curious where this timeout value came from. Clearly it works currently, but I wonder if this can fail if we add more `CellResponse` recordings in the future.

File: hnn_core/mpi_child.py
Comment (ntolley) [inline hnn_core/mpi_child.py:57]: Is the goal here to disentangle HNN vocabulary from generic MPI processes? I don't see an immediate use for this outside of passing the `Network` around. Not a big issue though considering the calls to this function are pretty transparent.

**Review by ntolley — APPROVED** [review]
Just to make sure I understand, the major changes were: 
1) moving the member functions outside of the backends class
2) Passing the `Network` to the child processes

The changes seem good to me but I'm certainly no authority on parallelization. @blakecaldwell may have some insight if there are potential issues that could crop up with passing larger more complex objects to processes. 

I'm happy to merge if you're satisfied, unless you want to hold out on a review from the expert ;)

**Comment by blakecaldwell** [comment]
Hi all. Thanks @jasmainak for the effort here. I'm happy to review if it can wait about 1 week from now.

**Comment by jasmainak** [comment]
sure! no rush on this PR :)

File: hnn_core/mpi_child.py
Comment (jasmainak) [reply hnn_core/mpi_child.py:57]: yes exactly ... if we can make things a bit more general/reusable, we can document and test them independently of the simulations. I don't think I achieved that though :)

File: hnn_core/parallel_backends.py
Comment (blakecaldwell) [reply hnn_core/parallel_backends.py:68]: I don't foresee any. We don't have users install openmpi directly, so are somewhat dependent on the versions that work best for each OS (installed by conda or Microsoft MPI).

File: doc/parallel.rst
Comment (blakecaldwell) [inline doc/parallel.rst:82]: Update to mention new vocabulary about objects being passed, though still technically accurate for this PR

File: hnn_core/parallel_backends.py
Comment (blakecaldwell) [reply hnn_core/parallel_backends.py:504]: It's not related to how much data is being passed. That is handled by the "end_of_data" signal. This just waits long enough to catch error messages if the process still has some in its stderr buffer. Also, it's unlikely that there are any errors at this point, during the shutdown process, but this just covers that case.

**Review by blakecaldwell — APPROVED** [review]
Looks good to me. It's hard to follow the changes made to handling the callback return values because "_read_stderr" and "_read_stdout" were moved outside of the classes, but I didn't see functional differences there.

I think it's a solid improvement @jasmainak!

**Comment by blakecaldwell** [comment]
@jasmainak what do you think is a path forward here? some of these changes are already part of master (e.g. passing `net` instead of `params`).

**Comment by jasmainak** [comment]
yeah, not sure if it's worth getting back to this PR. Specially since you're planning to change it to the `Queue` system. I'm fine closing this pull request.

**Comment by jasmainak** [comment]
closed by #296

**Pull request closed without merging**
